firmware/system_CAN.c

```diff
@@ -63,109 +96,82 @@ static void _dispatch_ctrl_rx(CANRxFrame *rx_msg)
 
     uint8_t ctrl_cmd = rx_msg->data8[0];
     switch(ctrl_cmd) {
-        case CTRL_CMD_RESET_STN1110:
-        {
-            uint8_t protocol = rx_msg->data8[1];
-            stn1110_reset(protocol);
-            break;
-        }
-        default:
-            log_info(LOG_PFX "Unknown control message command: %i\r\n", ctrl_cmd);
-            break;
+    case CTRL_CMD_CONFIGURE: {
+        _process_configure_cmd(rx_msg);
+        break;
     }
-}
-
-static void _log_CAN_rx_message(CANRxFrame * can_frame)
-{
-    if (get_logging_level() < logging_level_trace)
-        return;
-
-    uint32_t CAN_id = can_frame->IDE == CAN_IDE_EXT ? can_frame->EID : can_frame->SID;
-    log_trace(LOG_PFX "CAN Rx ID(%i): ", CAN_id);
-    size_t i;
-    for (i = 0; i < can_frame->DLC; i++)
-    {
-        log_trace("%02X ", can_frame->data8[i]);
+    default:
+        log_info(LOG_PFX "Unknown control message command: %i\r\n", ctrl_cmd);
+        break;
     }
-    log_trace("\r\n");
 }
 
-
+/* Process an incoming OBDII PID request */
 static void _process_pid_request(CANRxFrame *rx_msg)
 {
-    log_info(LOG_PFX "PID request\r\n");
-
-    if (!get_system_initialized())
-        return;
 
-    if (get_pid_request_active()){
-        log_info(LOG_PFX "ignoring, pid request active\r\n");
+    if (!get_system_initialized()) {
+        log_trace(LOG_PFX "Ignoring PID request: system initializing\r\n");
         return;
     }
 
     uint8_t data_byte_count = rx_msg->data8[0];
-    if (data_byte_count > MAX_PID_DATA_BYTES) {
-        log_info(LOG_PFX "Invalid PID request; max data bytes %i exceeded %i\r\n", data_byte_count, MAX_PID_DATA_BYTES);
+    if (data_byte_count > MAX_CAN_MESSAGE_SIZE - 1) {
+        log_info(LOG_PFX "Invalid PID request; max data bytes %i exceeded %i\r\n", data_byte_count, MAX_CAN_MESSAGE_SIZE - 1);
         return;
     }
-    size_t i;
-    for (i = 0; i < data_byte_count; i++) {
-        chprintf((BaseSequentialStream *)&SD2, "%02X", rx_msg->data8[i + 1]);
-    }
-    chprintf((BaseSequentialStream *)&SD2, "\r");
-    set_pid_request_active(true);
+
+    send_stn1110_pid_request(rx_msg->data8 + 1, data_byte_count);
 }
 
 /*
  * Dispatch an incoming CAN message
  */
 void dispatch_can_rx(CANRxFrame *rx_msg)
 {
-    /* we are only handling extended IDs */
-
     uint8_t can_id_type = rx_msg->IDE;
 
     switch (can_id_type) {
-        case CAN_IDE_EXT:
+    case CAN_IDE_EXT:
         /* Process Extended CAN IDs */
-        {
-            switch (rx_msg->EID){
-                case OBD2CAN_CTRL_ID:
-                    _dispatch_ctrl_rx(rx_msg);
-                    break;
-            }
+    {
+        switch (rx_msg->EID) {
+        case OBD2CAN_CTRL_ID:
```

Comment:
on the surface, yes, but it provides a framework for adding more control IDs in the future. that's why it's put in here explicitly .